wave_lang/kernel/wave/water_emitter.py

```diff
@@ -0,0 +1,364 @@
+"""
+
+import json
+import sys
+
+try:
+    from water_mlir import ir
+    from water_mlir.water_mlir.dialects.wave import (
+        AddOp,
+        DivOp,
+        Exp2Op,
+        MulOp,
+        ReadOp,
+        RegisterOp,
+        WriteOp,
+    )
+    from water_mlir.water_mlir.dialects import wave
+except Exception as e:
+    print(f"FATAL: failed to import water_mlir: {e}", file=sys.stderr)
+    sys.exit(2)
+
+# Mapping from tkw_op_name to actual op constructors
+WAVE_OP_CONSTRUCTORS = {
+    "add": AddOp,
+    "mul": MulOp,
+    "div": DivOp,
+    "exp2": Exp2Op,
+    "read": ReadOp,
+    "write": WriteOp,
+    "register": RegisterOp,
+    # TODO: Add more or find a good way of avoiding needing a mapping
+}
+
+
+def _default_tensor_type():
+    # Return a default Wave tensor type for register operations
+    return ir.Type.parse("!wave.tensor<any of f32, <register>>")
+
+
+def _default_memory_tensor_type():
+    # Return a default Wave tensor type for memory operations
+    return ir.Type.parse("!wave.tensor<any of f32, <global>>")
+
+
+def _map_address_space(addr: str) -> str:
+    if not addr:
+        return ""
+    a = str(addr).lower()
+    if "global" in a:
+        return "global"
+    if "shared" in a:
+        return "shared"
+    if "register" in a:
+        return "register"
+    # If we can't determine the address space, use unspecified
+    return "unspecified"
+
+
+def _build_wave_type_text(wave_type_dict: dict) -> str:
+    kind = wave_type_dict.get("kind")
+    shape = wave_type_dict.get("shape", []) or []
+    dtype = wave_type_dict.get("dtype", "f32")
+    addr = wave_type_dict.get("address_space", "")
+    if not shape:
+        shape_txt = "any"
+    else:
+        shape_txt = ", ".join([f"@{s}" for s in shape])
+    addr_norm = _map_address_space(addr)
+    addr_txt = f", <{addr_norm}>" if addr_norm else ""
+    return f"!wave.tensor<[{shape_txt}] of {dtype}{addr_txt}>"
+
+
+def _type_from_wave_type_dict(ctx: ir.Context, wave_type_dict: dict) -> ir.Type:
+    # Use textual parser to construct the Wave type.
+    type_text = _build_wave_type_text(wave_type_dict)
+    return ir.Type.parse(type_text, context=ctx)
+
+
+def create_operation_attributes(node_data):
+    """Create MLIR attributes from node data."""
+    attrs = {}
+
+    # Map common attributes
+    if "name" in node_data:
+        attrs["fx.name"] = ir.StringAttr.get(str(node_data["name"]))
+    if "op" in node_data:
+        attrs["fx.op"] = ir.StringAttr.get(str(node_data["op"]))
+    if "tkw_op_name" in node_data:
+        attrs["fx.tkw_op_name"] = ir.StringAttr.get(str(node_data["tkw_op_name"]))
+    if "subgraph" in node_data:
+        attrs["fx.subgraph"] = ir.StringAttr.get(str(node_data["subgraph"]))
+
+    # Map Python-specific attributes
+    for py_attr in (
+        "vector_shapes",
+        "reduction_dim",
+        "iter_idx",
+        "location",
+        "expanded_dims",
+        "scheduling_parameters",
+    ):
```

Comment:
Can we just have a python dataclass for these defined in a separate file that both this and `mlir_converter` import? Such a file will not be allowed to import anything beyond builtins. We can then serialize or even pickle such objects for communication.